### PR TITLE
update models.Vgg16

### PR DESCRIPTION
**Pull request by zsdonghao**

### CODE REVIEW

File: tensorlayer/models/vgg16.py [hunk 1]
```diff
@@ -158,54 +173,173 @@ def conv_layers_simple_api(net_in):
...
+            self.net = VGG16Base.vgg16_simple_api(net, end_with)
+            self.outputs = self.net.outputs
+            self.all_params = self.net.all_params
+            self.all_layers = self.net.all_layers
+            self.all_drop = self.net.all_drop
+            self.print_layers = self.net.print_layers
+            self.print_params = self.net.print_params
+
+
+# class VGG16(VGG16Base):
```
Comment (luomai) [inline tensorlayer/models/vgg16.py]: Remove unused code.

File: tensorlayer/models/vgg16.py [hunk 2]
```diff
@@ -158,54 +173,173 @@ def conv_layers_simple_api(net_in):
             net_in.outputs = net_in.outputs - mean
 
         # conv1
-        network = Conv2d(net_in, n_filter=64, filter_size=(3, 3), strides=(1, 1), act=tf.nn.relu, padding='SAME', name='conv1_1')
-        network = Conv2d(network, n_filter=64, filter_size=(3, 3), strides=(1, 1), act=tf.nn.relu, padding='SAME', name='conv1_2')
-        network = MaxPool2d(network, filter_size=(2, 2), strides=(2, 2), padding='SAME', name='pool1')
+        net = Conv2d(net_in, n_filter=64, filter_size=(3, 3), strides=(1, 1), act=tf.nn.relu, padding='SAME', name='conv1_1')
+        if end_with in net.name:
```
Comment (luomai) [inline tensorlayer/models/vgg16.py]: Shall we use "is" instead of "in"?

File: tensorlayer/models/vgg16.py [hunk 3]
```diff
@@ -158,54 +173,173 @@ def conv_layers_simple_api(net_in):
...
+            return net
+        net = Conv2d(net, n_filter=64, filter_size=(3, 3), strides=(1, 1), act=tf.nn.relu, padding='SAME', name='conv1_2')
+        if end_with in net.name:
```
Comment (luomai) [inline tensorlayer/models/vgg16.py]: The way we check early return is very repetitive. Is there a better way to doing this?

File: tensorlayer/models/vgg16.py (hunk 1 above)
Comment (zsdonghao) [reply tensorlayer/models/vgg16.py]: ok

File: tensorlayer/models/vgg16.py (hunk 2 above)
Comment (zsdonghao) [reply tensorlayer/models/vgg16.py]: we need to use `in`, because we are using `with tf.variable_scope` ~

File: tensorlayer/models/vgg16.py (hunk 3 above)
Comment (zsdonghao) [reply tensorlayer/models/vgg16.py]: no idea now

Comment (lgarithm) [reply tensorlayer/models/vgg16.py]: we need to introduce something that can implement **layer as unary operator**.

Comment (lgarithm) [reply tensorlayer/models/vgg16.py]: e.g. https://github.com/tensorlayer/tensorlayer/commit/734aa01c5495ba95efdb776de3a9f6640c9ac5ff

File: tensorlayer/models/vgg16.py [hunk 4]
```diff
@@ -42,170 +49,244 @@ def conv_layers(net_in):
...
+    >>> net = tl.layers.DenseLayer(vgg, 100, name='out')
+    >>> # initialize all parameters
+    >>> sess = tf.InteractiveSession()
+    >>> tl.layers.initialize_global_variables(sess)
+    >>> # restore pre-trained VGG parameters
+    >>> vgg.restore_params(sess)
+    >>> # train your own classifier (only update the last layer)
+    >>> train_params = tl.layers.get_variables_with_name('out')
+
+    - Reuse model
```
Comment (luomai) [inline tensorlayer/models/vgg16.py]: remove listing

File: tensorlayer/models/vgg16.py [hunk 5]
```diff
@@ -42,170 +49,244 @@ def conv_layers(net_in):
...
+    >>> x = tf.placeholder(tf.float32, [None, 224, 224, 3])
+    >>> # get the whole model
+    >>> vgg = tl.models.VGG16(x)
+    >>> # restore pre-trained VGG parameters
+    >>> sess = tf.InteractiveSession()
+    >>> vgg.restore_params(sess)
+    >>> # use for inferencing
+    >>> probs = tf.nn.softmax(vgg.outputs)
+
+    - Extract features with VGG16 and Train a classifier with 100 classes
```
Comment (luomai) [inline tensorlayer/models/vgg16.py]: remove listing

File: tensorlayer/models/vgg16.py [hunk 6]
```diff
@@ -42,170 +49,244 @@ def conv_layers(net_in):
...
+    x : placeholder
+        shape [None, 224, 224, 3], value range [0, 255].
+    end_with : str
+        The end point of the model. Default ``fc3_relu`` i.e. the whole model.
+    reuse : boolean
+        Whether to reuse the model.
+
+    Examples
+    ---------
+    - Classify ImageNet classes with VGG16, see `tutorial_models_vgg16.py <https://github.com/tensorlayer/tensorlayer/blob/master/example/tutorial_models_vgg16.py>__`
```
Comment (luomai) [inline tensorlayer/models/vgg16.py]: remove listing

**Pull request merged**